pyomo/contrib/satsolver/satsolver.py

```diff
@@ -0,0 +1,273 @@
+                                         AbsExpression,
+                                         UnaryFunctionExpression,
+                                         nonpyomo_leaf_types,
+                                         StreamBasedExpressionVisitor)
+from pyomo.environ import SymbolMap, NumericLabeler, Var, Constraint
+from pyomo.gdp import Disjunction
+
+_z3_available = True
+try:
+    import z3
+except ImportError:
+    _z3_available = False
+
+
+def satisfiable(model):
+    """Checks if the model is satisfiable.
+
+    Returns:
+        True  if model is satisfiable,
+        False if model is unsatisfiable,
+        None  if satisfiability cannot be determined.
+    """
+    result = SMTSatSolver(model).check()
+    if result == z3.sat:
+        return True
+    elif result == z3.unsat:
+        return False
+    elif result == z3.unknown:
+        return None
+    else:
+        raise ValueError('Unknown result: %s' % result)
+
+
+class SMTSatSolver(object):
+    """
+    Satisfiability solver that checks constraint feasibility through use of
+    z3 Sat Solver. Object stores expressions and variables in form consistent
+    with SMT-LIB standard.
+    For documentation on SMT-LIB standard see
+    http://smtlib.cs.uiowa.edu/
+    """
+
+    def __str__(self):
+        """
+        Defined string representation of object
+        """
+        string = ""
+        string = string + "Variables:\n"
+        for v in self.variable_list:
+            string = string + v
+        string = string + "Bounds:\n"
+        for e in self.bounds_list:
+            string = string + e
+        string = string + "Expressions:\n"
+        for e in self.expression_list:
+            string = string + e
+        string = string + "Disjunctions:\n"
+        for djn in self.disjunctions_list:
+            string = string + "Disjunction: " + djn[0] + "\n"
+            for disj in djn[1]:
+                string = string + "  " + disj[0] + " : " + "\n"
+                for c in disj[1]:
+                    string = string + "    " + c + "\n"
+        return string
+
+    def __init__(self, model=None):
+        self.variable_label_map = SymbolMap(NumericLabeler('x'))
+        self.prefix_expr_list = self._get_default_functions()
+        self.variable_list = []
+        self.bounds_list = []
+        self.expression_list = []
+        self.disjunctions_list = []
+        self.walker = SMT_visitor(self.variable_label_map)
+        self.solver = z3.Solver()
+
+        if model is not None:
+            self._process_model(model)
+
+    # Set up functions to be added to beginning of string
+    def _get_default_functions(self):
+        default = list()
+        default.append("(define-fun exp ((x Real)) Real (^ 2.718281828459045 x))")
+        return default
+
+    # processes pyomo model into SMT model
+    def _process_model(self, model):
+        for v in model.component_data_objects(ctype=Var, descend_into=False):
```

Comment:
This looks like it only handles "flat" models with Disjunctions.  Sort of.  You will only pick Vars on the top-level model, but you will pick up Constraints in any active Block.  Is that the desired behavior?

Reply:
Thanks for the catch, believe it was left in from debugging.